### PR TITLE
Remove instance level streams lock

### LINKED ISSUE #5177: Performance: Push stuck at ingester's streamsMtx.Lock
**Is your feature request related to a problem? Please describe.**
I've noticed some performance issues for my Loki clusters after upgrading to `2.4.1`(exist before `2.4.1`, but discovered after upgrading).
About 1% of push requests (≈60 QPS) return code `500` with `context canceled` message after 2 seconds timeout.
Tracing data shows these failed requests spend too much time(6s~8s) on ingesters' `Push` GRPC calls, distributors return before ingesters return due to configured 2s timeout.
After diagnosing with pprof, I found out most these slow Push calls are stuck at 

https://github.com/grafana/loki/blob/985e965796c7c95c5c2999bc5a212313fc2a673a/pkg/ingester/instance.go#L157

which is an instance level streams lock.

This lock limits every Push calls to execute one at a time, causing performance bottleneck.

**Describe the solution you'd like**
Remove this instance level lock, make streams/streamsByFP concurrently readable.

**Describe alternatives you've considered**
None

**Additional context**

### Hardware

All targets are deployed directly to individual physical machines.

| Target | Instances num | CPU | Ram | Ethernet |
|-|-|-|-|-|
| Distributor | 2 | Intel(R) Xeon(R) Silver 4210 @ 2.20GHz * 2 | DDR4 16GB * 8 | 10Gb Fiber |
| Ingester | 5 | Intel(R) Xeon(R) Silver 4210 @ 2.20GHz * 2 | DDR4 32GB * 8 | 10Gb Fiber |

Other unrelated targets are not listed.

### Cluster Status

| Metric | Value | Prometheus metric |
|-|-|-|
| Distributor Throughput (Total) | ≈600 MiB/s | sum(irate(loki_distributor_bytes_received_total{}[2m])) |
| Ingester Memory Streams (Total) | ≈9000 | sum(loki_ingester_memory_streams{}) |
| Push QPS (Total) | ≈6000 | sum(rate(loki_request_duration_seconds_count{method=~"POST"}[2m])) |

### Log

```
level=warn ts=2022-01-19T01:59:27.715023461Z caller=logging.go:72 orgID=xxxxxx msg="POST /loki/api/v1/push (500) 1.999633256s Response: \"context canceled\\n\" ws: false; Connection: close; Content-Length: 53357; Content-Type: application/x-protobuf; User-Agent: promtail/; X-Nginx-Proxy: true; X-Original-Uri: /api/v1/loki/api/v1/push; X-Real-Ip: 10.xxx.xxx.xxx; X-Request-Id: c32c9d2cbedc22b1c2b73b09c2b7fc50; X-Scheme: http; X-Scope-Orgid: xxxxxx; "
```
Sensitive information is censored.

### Tracing

Tracing data shows these failed requests spend too much time(6s~8s) on ingesters' `Push` GRPC calls, distributors return before ingesters return due to configured 2s timeout.

skywalking

After checking several failed requests, all ingester instances appear to have this issue irregularly.

### PProf

Since tracing data shows this is an ingester issue, I've managed to capture goroutine info with pprof.

```
goroutine profile: total 886
631 @ 0x43b285 0x44cfc5 0x44cfae 0x46e767 0x48d625 0x48ed90 0x48ed22 0x22ba093 0x22b6406 0x191d849 0x229f1c2 0xa35d83 0x1ced756 0xa35d83 0x1e0178b 0xa35d83 0xb72464 0x209acd6 0xa35d83 0xb7585a 0xa35d83 0xa3a062 0xa35d83 0xb72a14 0xa35d83 0xa35f77 0x18f6a10 0x9ddaab 0x9e1f4c 0x9f0ecb 0x472881
#	0x46e766	sync.runtime_SemacquireMutex+0x46										/go/src/runtime/sema.go:71
#	0x48d624	sync.(*Mutex).lockSlow+0x104											/go/src/sync/mutex.go:138
#	0x48ed8f	sync.(*Mutex).Lock+0x8f												/go/src/sync/mutex.go:81
#	0x48ed21	sync.(*RWMutex).Lock+0x21											/go/src/sync/rwmutex.go:111
#	0x22ba092	github.com/grafana/loki/pkg/ingester.(*instance).Push+0x152							/buildspace/e4e619d835/loki/pkg/ingester/instance.go:161
#	0x22b6405	github.com/grafana/loki/pkg/ingester.(*Ingester).Push+0xe5							/buildspace/e4e619d835/loki/pkg/ingester/ingester.go:503
#	0x191d848	github.com/grafana/loki/pkg/logproto._Pusher_Push_Handler.func1+0x88						/buildspace/e4e619d835/loki/pkg/logproto/logproto.pb.go:2727
...
```

### PR DESCRIPTION
**Pull request by RangerCD**
**What this PR does / why we need it**:

Remove instance level `streams`/`streamsByFP` lock inside ingester, `Push` GRPC calls can run parallel to others with different streams.

This PR introduces a new struct `streamsMap` to replace `streamsMtx`, `streams`, `streamsByFP` inside ingester's `instance`. Instead of locking `streams`/`streamsByFP` while reading/writing, this struct provides `Load`/`LoadByFP` which is basically lock-free to read from internal `sync.Map`. Writing to `streamsMap` is locked internally to keep consistency of two `sync.Map`s.

Significantly improve concurrent capacity and process speed of ingester's `Push` GRPC API, as well as distributor's `/loki/api/v1/push` HTTP API. 

**Which issue(s) this PR fixes**:

Fixes #5177

**Special notes for your reviewer**:

This PR was tested based on Loki `2.4.1`, due to our production environment baseline version is Loki `2.4.1`, there is no related commits since `2.4.1` on `main` branch until now.

This PR has been tested over several clusters, charts bellow are from the same cluster mentioned in #5177

loki_qps

Total QPS has less spikes with this PR. POST 500 rate also dropped from 1%(QPS≈60) to 0.005%(QPS≈0.3)

throughput_per_tenant

Distributor throughput is also smoother than before.

grpc_duration

Push GRPC API request duration now has a more stable and satisfying value.

post_duration

Push HTTP POST API has similar performance improvement.

**Checklist**
- [ ] Documentation added
- [x] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

**Comment by cyriltovena** [comment]
The kind of PR that reminds me why I work in the OSS.

**Comment by RangerCD** [comment]
Another commit to solve Load and Delete consistency problem is being tested currently.

**Review by sandeepsukhani — APPROVED** [review]
Nice job! Changes look great to me!
Thank you so much for taking time to improve the code!

File: pkg/ingester/streams_map.go
Comment (cyriltovena) [inline pkg/ingester/streams_map.go:11]: nit: Can you add a comment that this is for keeping the consistency between the two map please ?

**Review by chaudum — APPROVED** [review]
This looks great. Thanks for the contribution!

**Review by owen-d — APPROVED** [review]
Wonderful PR; this made my morning.

**Comment by RangerCD** [comment]
> This is fantastic! Thank you!
> 
> > Another commit to solve Load and Delete consistency problem is being tested currently.
> 
> Should we wait to merge this PR? Or will that be a separate PR?

Please wait for that commit, I'll push it to this PR. It's necessary to let it run & test for longer period since streams is really important to ingester.
This Load & Delete conflict would happen when Push loads a stream while it is being removed by scheduled flush. It's a low probability event, but it is possible. Which means this PR is not complete without solving this problem.

File: pkg/ingester/streams_map.go
Comment (RangerCD) [reply pkg/ingester/streams_map.go:62]: That sounds reasonable. But maybe returns a `[]error` could be better? If `fn` returns `true` and an `error`, continue and collect this `error`. `true` but actually stops iteration is kind of counterintuitive to me.

File: pkg/ingester/streams_map_test.go
Comment (trevorwhitney) [inline pkg/ingester/streams_map_test.go:69]: super tiny nit: since this test is making sure both the `LoadOrStoreNewByFP` calls worked, and thus not directly related to `len(ss)`, I would test against the hard-coded 2 here just so the reader doesn't have to keep that extra context of how big `ss` is.

```suggestion
	require.Equal(t, 2, streams.Len())
```

**Review by trevorwhitney — COMMENTED** [review]
This is really great, thanks for the contribution! I have the tiniest of nits to one of the test assertions, take it or leave it, your call :)

File: pkg/ingester/streams_map_test.go
Comment (owen-d) [reply pkg/ingester/streams_map_test.go:69]: I believe this makes more sense as it is, because we can safely add more streams to the test case and this will reflect the expectation that they're all stored. The prior checks could more idiomatically be run in a loop over each stream, but that's a tiny suggestion.

Comment (trevorwhitney) [reply pkg/ingester/streams_map_test.go:69]: yeah, I guess my thought is that adding more cases to the slice won't actually cause them to be stored, because it's a function of actually calling the `LoadOrStore`. However, if we put that in a loop like you mentioned then I can see this argument. But now we're splitting hairs, doesn't really matter either way.

**Comment by RangerCD** [comment]
Another commit has been pushed to this PR, to solve Load and Delete consistency problem.
Review suggestions will be added to next commit.

## Problem

Periodic flush will try to remove empty stream at the end of `removeFlushedChunks`. If a stream loaded from `streamsMap`, while it's being flushed, there is a small chance this stream will be removed and `stream.Push` is still trying to write. To produce this problem, there are several requirements:

1. This stream will be empty after current flush
2. A `Push` request contains the same `Labels` comes in right before `removeFlushedChunks`
3. `LoadOrStoreNew` returns this stream, but its `chunkMtx` locked by `removeFlushedChunks` right before `stream.Push` doing so

In this case, current `Push` will write to a stream that is no longer exist in `streamsMap`, causing data loss.

## Solution

Double check this stream with `chunkMtx` locked, hold this lock until `stream.Push` finishes. If the stream is newly created by `LoadOrStoreNew`, lock its `chunkMtx` before it has been stored to `streamsMap`.

## Implementation

This commit adds a new function `executeWithGetOrCreateStream`(maybe a better name?), it's basically a safer version of `getOrCreateStream`. The difference between these two function is the new one requires `fn func(*stream) error` as additional parameter. `executeWithGetOrCreateStream` will try to lock stream's `chunkMtx`, execute `fn` with this lock being held, it will always pass a valid stream to `fn`. Some `getOrCreateStream` use cases could still remain using it, if it's not sensitive to Load & Delete consistency. For example, `ingesterRecoverer` doesn't care about deletion because flush loop is even not started yet.

## Performance

No noticeable performance change for this commit.

File: pkg/ingester/instance.go
Comment (owen-d) [inline pkg/ingester/instance.go:263]: It may be more intuitive to provide a helper function which will get/create a stream and execute a function under one lock acquisition. In addition to reducing the streams lock contention by 50% (no need to double check after the first create), it will present a simpler interface. Something like:
```
// Executes a function on a stream with the lock held.
// Will create the stream according to `newStreamFn` if it doesn't exist
func (m *streamsMap) UnderLock(
	pushReqStream logproto.Stream,
	newStreamFn func() (*stream, error),
	fn func(*stream) error,
	) (error) {
	// lock
	// get or create
	// run fn on stream
	// unlock
}
```

**Review by owen-d — COMMENTED** [review]
I believe there's a new very unlikely dataloss bug if we merge the PR as is. Consider the case:

```
1) Stream goes idle
2) We begin flushing the stream, acquiring the chunkMtx: https://github.com/grafana/loki/blob/main/pkg/ingester/flush.go#L311
3) A new push comes in for stream, resolves to the currently flushing `*stream` and begins waiting to acquire chunkMtx: https://github.com/grafana/loki/blob/main/pkg/ingester/stream.go#L180
4) Flush removes the stream: https://github.com/grafana/loki/blob/main/pkg/ingester/flush.go#L330 and unlocks
5) The stream is removed from the `*instance`
6) The now orphaned `*stream` has it's mutex acquired by the `Push()` call, writes to the orphaned stream, and exits
7) Orphaned stream is garbage collected
8) Data loss :(
```

File: pkg/ingester/instance.go
Comment (RangerCD) [reply pkg/ingester/instance.go:263]: The lock here is `chunkMtx` under `stream`, instead of `consistencyMtx` under `streamsMap`. Which means we have to get the stream first, then lock the `chunkMtx` under it. The reason to lock `chunkMtx` is to make sure `stream` is not being deleted by `removeFlushedChunks` while `fn` is running.

**Comment by owen-d** [comment]
Good point!

I suspect this will still be an overall benefit for reads, but it's probably worth rerunning some tests.

Previously, pushes would:
1) lock streams mtx
2) lock chunk mtx
3) push
4) unlock chunk mtx
5) unlock streams mtx

The locks were held for the duration of the push, totaling one lock of each `streamsMtx`, `chunksMtx`

Now, pushes:
1) lock consistency mtx
2) get|create stream
3) unlock consistency mtx
4) lock chunk mtx
5) lock consistency mtx
6) re-fetch stream to ensure we have the correct pointer
7) unlock consistency mtx
8) unlock chunk mtx

Previously, we locked two distinct mutexes one time each. Now, we lock two distinct mutexes, one twice and one once. However, we won't actually be holding anything but the chunkMtx when the actual `Push` is executed, which will still free up lock contention for reads dramatically. Nice work.

I think we can improve this even further with a bit of refactoring. If `removeFlushedChunks` locked the consistency mtx as well, we wouldn't need to re-lock the consistency mtx when pushing to ensure we have the correct stream pointer.

Instead, we could have
Push: `lock consistency mtx -> lock chunks mtx -> unlock consistency mtx -> push -> unlock chunks mtx`
Flush: `lock consistency mtx -> lock chunks mtx -> flush + remove stream -> unlock chunks mtx -> unlock consistency mtx`

I think this would allow us to reduce the number of lock calls on the push path at the expense of some additional locking on the flush path which is run much less often. Just an idea, let me know if you see issues with it and we don't have to put it in this PR either.

**Comment by RangerCD** [comment]
> I think we can improve this even further with a bit of refactoring. If `removeFlushedChunks` locked the consistency mtx as well, we wouldn't need to re-lock the consistency mtx when pushing to ensure we have the correct stream pointer.
> 
> Instead, we could have Push: `lock consistency mtx -> lock chunks mtx -> unlock consistency mtx -> push -> unlock chunks mtx` Flush: `lock consistency mtx -> lock chunks mtx -> flush + remove stream -> unlock chunks mtx -> unlock consistency mtx`
> 
> I think this would allow us to reduce the number of lock calls on the push path at the expense of some additional locking on the flush path which is run much less often. Just an idea, let me know if you see issues with it and we don't have to put it in this PR either.

This idea is absolutely worth a try! It looks good to me, and it would not be hard to implement.
I'm curious about its performance, it might take several days to test and run for a final conclusion. I'd like to follow your opinion of whether to put it to this PR or another one.

**Review by owen-d — APPROVED** [review]
I'll merge this first; we can do a followup PR with the remaining suggestions. 

Great work. I'll merge later today in the case there's no more feedback from others.

**Comment by owen-d** [comment]
Would you like to do the followup PR or should I @RangerCD?

**Comment by RangerCD** [comment]
> Would you like to do the followup PR or should I @RangerCD?

I'm willing to do so. It will come next week, I have some other works piled up to finish first.@owen-d

File: pkg/ingester/streams_map.go
Comment (RangerCD) [reply pkg/ingester/streams_map.go:11]: The comment commit was missing from this PR, I'll fill it up in next one.

**Pull request merged**
